### DISCUSSION TITLE
Build error in "Linking CXX executable struc_to_unstruc"

### DISCUSSION THREAD

**Discussion by erdchris** [sandialabs/seacas#417] [Q&A]
Hello. I'm trying to build the latest SEACAS under RHEL 8.8 using Intel 2022.2.0 (and GCC 8.2.0). Build dies here:

[ 43%] Linking CXX executable struc_to_unstruc
CMakeFiles/struc_to_unstruc.dir/struc_to_unstruc.C.o: In function `main':
struc_to_unstruc.C:(.text+0x97): undefined reference to `fmt::v10::vprint(_IO_FILE*, fmt::v10::basic_string_view<char>, fmt::v10::basic_format_args<fmt::v10::basic_format_context<fmt::v10::appender, char> >)'
struc_to_unstruc.C:(.text+0x271): undefined reference to `fmt::v10::vprint(_IO_FILE*, fmt::v10::basic_string_view<char>, fmt::v10::basic_format_args<fmt::v10::basic_format_context<fmt::v10::appender, char> >)'
struc_to_unstruc.C:(.text+0x4c1): undefined reference to `fmt::v10::vprint(_IO_FILE*, fmt::v10::basic_string_view<char>, fmt::v10::basic_format_args<fmt::v10::basic_format_context<fmt::v10::appender, char> >)'
struc_to_unstruc.C:(.text+0x504): undefined reference to `fmt::v10::vprint(_IO_FILE*, fmt::v10::basic_string_view<char>, fmt::v10::basic_format_args<fmt::v10::basic_format_context<fmt::v10::appender, char> >)'
CMakeFiles/struc_to_unstruc.dir/struc_to_unstruc.C.o: In function `(anonymous namespace)::create_unstructured(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
struc_to_unstruc.C:(.text+0x866): undefined reference to `fmt::v10::vprint(_IO_FILE*, fmt::v10::basic_string_view<char>, fmt::v10::basic_format_args<fmt::v10::basic_format_context<fmt::v10::appender, char> >)'
CMakeFiles/struc_to_unstruc.dir/struc_to_unstruc.C.o:struc_to_unstruc.C:(.text+0xb7b): more undefined references to `fmt::v10::vprint(_IO_FILE*, fmt::v10::basic_string_view<char>, fmt::v10::basic_format_args<fmt::v10::basic_format_context<fmt::v10::appender, char> >)' follow
CMakeFiles/struc_to_unstruc.dir/struc_to_unstruc.C.o: In function `(anonymous namespace)::create_unstructured(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
struc_to_unstruc.C:(.text+0xedb): undefined reference to `fmt::v10::vprint(fmt::v10::basic_string_view<char>, fmt::v10::basic_format_args<fmt::v10::basic_format_context<fmt::v10::appender, char> >)'
struc_to_unstruc.C:(.text+0x1014): undefined reference to `fmt::v10::vprint(_IO_FILE*, fmt::v10::basic_string_view<char>, fmt::v10::basic_format_args<fmt::v10::basic_format_context<fmt::v10::appender, char> >)'
struc_to_unstruc.C:(.text+0x140e): undefined reference to `fmt::v10::vprint(_IO_FILE*, fmt::v10::basic_string_view<char>, fmt::v10::basic_format_args<fmt::v10::basic_format_context<fmt::v10::appender, char> >)'
CMakeFiles/struc_to_unstruc.dir/struc_to_unstruc.C.o: In function `(anonymous namespace)::transfer_sidesets(Ioss::Region&, Ioss::Region&)':
struc_to_unstruc.C:(.text+0x4e92): undefined reference to `fmt::v10::vprint(fmt::v10::basic_string_view<char>, fmt::v10::basic_format_args<fmt::v10::basic_format_context<fmt::v10::appender, char> >)'
CMakeFiles/struc_to_unstruc.dir/struc_to_unstruc.C.o: In function `(anonymous namespace)::transfer_nodeblock(Ioss::Region&, Ioss::Region&)':
struc_to_unstruc.C:(.text+0x58a7): undefined reference to `fmt::v10::vprint(fmt::v10::basic_string_view<char>, fmt::v10::basic_format_args<fmt::v10::basic_format_context<fmt::v10::appender, char> >)'
struc_to_unstruc.C:(.text+0x58f1): undefined reference to `fmt::v10::vprint(fmt::v10::basic_string_view<char>, fmt::v10::basic_format_args<fmt::v10::basic_format_context<fmt::v10::appender, char> >)'
CMakeFiles/struc_to_unstruc.dir/struc_to_unstruc.C.o: In function `char const* fmt::v10::detail::parse_format_specs<char>(char const*, char const*, fmt::v10::detail::dynamic_format_specs<char>&, fmt::v10::basic_format_parse_context<char>&, fmt::v10::detail::type)':
struc_to_unstruc.C:(.text._ZN3fmt3v106detail18parse_format_specsIcEEPKT_S5_S5_RNS1_20dynamic_format_specsIS3_EERNS0_26basic_format_parse_contextIS3_EENS1_4typeE[_ZN3fmt3v106detail18parse_format_specsIcEEPKT_S5_S5_RNS1_20dynamic_format_specsIS3_EERNS0_26basic_format_parse_contextIS3_EENS1_4typeE]+0x951): undefined reference to `fmt::v10::detail::throw_format_error(char const*)'
struc_to_unstruc.C:(.text._ZN3fmt3v106detail18parse_format_specsIcEEPKT_S5_S5_RNS1_20dynamic_format_specsIS3_EERNS0_26basic_format_parse_contextIS3_EENS1_4typeE[_ZN3fmt3v106detail18parse_format_specsIcEEPKT_S5_S5_RNS1_20dynamic_format_specsIS3_EERNS0_26basic_format_parse_contextIS3_EENS1_4typeE]+0xcfe): undefined reference to `fmt::v10::detail::throw_format_error(char const*)'
struc_to_unstruc.C:(.text._ZN3fmt3v106detail18parse_format_specsIcEEPKT_S5_S5_RNS1_20dynamic_format_specsIS3_EERNS0_26basic_format_parse_contextIS3_EENS1_4typeE[_ZN3fmt3v106detail18parse_format_specsIcEEPKT_S5_S5_RNS1_20dynamic_format_specsIS3_EERNS0_26basic_format_parse_contextIS3_EENS1_4typeE]+0xe48): undefined reference to `fmt::v10::detail::throw_format_error(char const*)'
struc_to_unstruc.C:(.text._ZN3fmt3v106detail18parse_format_specsIcEEPKT_S5_S5_RNS1_20dynamic_format_specsIS3_EERNS0_26basic_format_parse_contextIS3_EENS1_4typeE[_ZN3fmt3v106detail18parse_format_specsIcEEPKT_S5_S5_RNS1_20dynamic_format_specsIS3_EERNS0_26basic_format_parse_contextIS3_EENS1_4typeE]+0xe6e): undefined reference to `fmt::v10::detail::throw_format_error(char const*)'
struc_to_unstruc.C:(.text._ZN3fmt3v106detail18parse_format_specsIcEEPKT_S5_S5_RNS1_20dynamic_format_specsIS3_EERNS0_26basic_format_parse_contextIS3_EENS1_4typeE[_ZN3fmt3v106detail18parse_format_specsIcEEPKT_S5_S5_RNS1_20dynamic_format_specsIS3_EERNS0_26basic_format_parse_contextIS3_EENS1_4typeE]+0xf91): undefined reference to `fmt::v10::detail::throw_format_error(char const*)'
CMakeFiles/struc_to_unstruc.dir/struc_to_unstruc.C.o:struc_to_unstruc.C:(.text._ZN3fmt3v106detail18parse_format_specsIcEEPKT_S5_S5_RNS1_20dynamic_format_specsIS3_EERNS0_26basic_format_parse_contextIS3_EENS1_4typeE[_ZN3fmt3v106detail18parse_format_specsIcEEPKT_S5_S5_RNS1_20dynamic_format_specsIS3_EERNS0_26basic_format_parse_contextIS3_EENS1_4typeE]+0x1044): more undefined references to `fmt::v10::detail::throw_format_error(char const*)' follow
libio_info_lib.so.2.0: undefined reference to `fmt::v10::detail::dragonbox::get_cached_power(int)'
libio_info_lib.so.2.0: undefined reference to `std::locale fmt::v10::detail::locale_ref::get<std::locale>() const'
libio_info_lib.so.2.0: undefined reference to `char fmt::v10::detail::decimal_point_impl<char>(fmt::v10::detail::locale_ref)'
libio_info_lib.so.2.0: undefined reference to `void fmt::v10::detail::vformat_to<char>(fmt::v10::detail::buffer<char>&, fmt::v10::basic_string_view<char>, fmt::v10::detail::vformat_args<char>::type, fmt::v10::detail::locale_ref)'
libio_info_lib.so.2.0: undefined reference to `fmt::v10::detail::is_printable(unsigned int)'
libio_info_lib.so.2.0: undefined reference to `fmt::v10::detail::dragonbox::decimal_fp<double> fmt::v10::detail::dragonbox::to_decimal<double>(double)'
libio_info_lib.so.2.0: undefined reference to `fmt::v10::detail::dragonbox::decimal_fp<float> fmt::v10::detail::dragonbox::to_decimal<float>(float)'
../cgns/libIocgns.so.2.0: undefined reference to `fmt::v10::vformat[abi:cxx11](fmt::v10::basic_string_view<char>, fmt::v10::basic_format_args<fmt::v10::basic_format_context<fmt::v10::appender, char> >)'
libio_info_lib.so.2.0: undefined reference to `fmt::v10::detail::thousands_sep_result<char> fmt::v10::detail::thousands_sep_impl<char>(fmt::v10::detail::locale_ref)'
libio_info_lib.so.2.0: undefined reference to `fmt::v10::detail::write_loc(fmt::v10::appender, fmt::v10::loc_value, fmt::v10::format_specs<char> const&, fmt::v10::detail::locale_ref)'
../libIoss.so.2.0: undefined reference to `fmt::v10::detail::print(_IO_FILE*, fmt::v10::basic_string_view<char>)'

**Comment by erdchris** [comment]
These were my steps:

```
export ACCESS=`pwd`
export SHARED=YES
export MPI=NO
export CRAY=NO
export COMPILER=intel
export INSTALL_PATH=/app/unsupported/SEACAS/v2023-11-04
mkdir $INSTALL_PATH

export CXX=icpc
export CC=icc
export FC=ifort
export F77=ifort
export F90=ifort
export CFLAGS="-O2 -fPIC -gcc-name=`which gcc`"
export CXXFLAGS="-O2 -fPIC -gxx-name=`which g++`"
export FFLAGS="-fPIC"
export LDFLAGS="-O2 -gcc-name=`which gcc` -gxx-name=`which g++`"

module load cmake/3.28.0-intel-2022.2.0

./install-tpl.sh

mkdir build
cd build
export BUILDDIR=`pwd`
../cmake-config
make
```

**Reply by gsjaardema** [reply]
It looks like lib::fmt was installed incorrectly, or was built with a different compiler or set of compiler flags.  Does `/app/unsupported/SEACAS/v2023-11-04/lib` have a `libfmt.so`?  

I don't know whether specifying both `COMPILER=intel` and the explicitly setting `CXX`, `CC`, ... is causing some confusion about versions...

**Reply by erdchris** [reply]
Thanks gsjaardema. The compiler and options should be the same for all parts. The lib does exist:

lib/libfmt.so -> libfmt.so.10
lib/libfmt.so.10 -> libfmt.so.10.1.1
-rwxr-xr-x. 1 user group 315024 Nov  6 08:54 lib/libfmt.so.10.1.1

I was using basically the same recipe from 2 years ago -- the last time I built SEACAS. I did build today with GNU only, and that worked.

**Reply by gsjaardema** [reply]
Ok,this might be related to intel then.  There have been issues in the past with libfmt and intel versions...

**Comment by gsjaardema** [comment]
I've moved this to an issue #422 since easier to track and I think more relavant...

**Comment by gsjaardema** [comment]
@erdchris  See temporary workaround/kluge in #422.
